### PR TITLE
Add --filter option to select components

### PR DESCRIPTION
**Pull request by txels**
Allows filtering based on any matching `key=value` in the component's `builder.ini` section.

### CODE REVIEW

**Comment by txels** [comment]
Ah works on 3.5 but fails on 2.7... will fix.

File: component-builder/tests/test_cli.py [hunk 1]
```diff
@@ -48,3 +47,34 @@ def test_discover_without_versions(self):
...
+            s.getvalue(),
+            '\n'.join([
+                'dummy-app',
+                'dummy-island-service',
+                ''
+            ])
+        )
+
+    @patch('sys.argv', ['compbuild', 'discover', '--all',
+                        '--filter=release-process=docker,label=app',
```
Comment (alexcouper) [inline component-builder/tests/test_cli.py:68]: nice!

File: component-builder/src/component_builder/discover.py [hunk 2]
```diff
@@ -13,15 +14,30 @@ def is_changed(candidate):
...
+        components = filter(partial(matches_selector, selector), components)
+
+    return list(components)
+
+
+def run(components, component_names=None, get_all=False, selectors=None):
     "Get paths and titles of changed components"
-    if label_filter and get_all:
+    if component_names and get_all:
         print("Asked to filter and get all. Assuming filter...")
```
Comment (alexcouper) [inline component-builder/src/component_builder/discover.py]: Maybe update this too - "asked for specific components and get all. Assuming specific components" ?

File: component-builder/src/component_builder/discover.py [hunk 3]
```diff
@@ -13,15 +14,30 @@ def is_changed(candidate):
...
 
-def run(components, label_filter=None, get_all=False):
+def filter_by(selectors, components):
+
+    def matches_selector(selector, component):
+        key, value = selector.split('=')
+        return component.ini.get(key) == value
+
+    for selector in selectors:
+        components = filter(partial(matches_selector, selector), components)
```
Comment (alexcouper) [inline component-builder/src/component_builder/discover.py:24]: Very elegantly done

**Comment by txels** [comment]
#preapproved-selfmerge

**Comment by txels** [comment]
Hmm I just realised that annoyingly this flag is only used when running discover, not build/test/etc...

**Pull request merged**